home_work/hw_10/type_decorator.py

```diff
@@ -0,0 +1,92 @@
+"""Module for type-based decorator."""
+from loguru import logger
+
+
+def typed(type_):
+    """Decorator factory to enforce argument types."""
+
+    def real_decorator(func):
+        """Real decorator to enforce argument types."""
+
+        def wrapper(*args, **kwargs):
+            """Wrapper function to enforce argument types."""
+            args = [arg if isinstance(arg, type_) else type_(arg) for arg in args]
+            kwargs = {k: v if isinstance(v, type_) else type_(v) for k, v in kwargs.items()}
+
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return real_decorator
+
+
+@typed(type_=str)
+def add_str(*args, **kwargs):
+    """Concatenate multiple strings."""
+    return ''.join(args) + ''.join(kwargs.values())
+
+
+@typed(type_=int)
+def add_int(*args, **kwargs):
+    """Add integers."""
+    return sum(args) + sum(kwargs.values())
+
+
+@typed(type_=float)
+def add_float(*args, **kwargs):
+    """Add floating point numbers."""
+    return sum(args) + sum(kwargs.values())
+
+
+assert add_str("3", 5) == '35', \
+    f"Expected '35', got {add_str('3', 5)}"
+
+assert add_str(5, 5) == '55', \
+    f"Expected '55', got {add_str(5, 5)}"
+
+assert add_str('a', 'b') == 'ab', \
+    f"Expected 'ab', got {add_str('a', 'b')}"
+
+assert add_int(5, "6", 7) == 18, \
+    f"Expected 18, got {add_int(5, 6, 7)}"
+
+assert add_float(0.1, 0.2, 0.4) == 0.7000000000000001, \
+    f"Expected 0.7000000000000001, got {add_float(0.1, 0.2, 0.4)}"
+
+
+##########################################################################
+# For Sergey Stamal
+# Why we should use decorator for typing if solution below seems more
+# flexible. Without uses any decorators.
```

Comment:
Good point, but in your example, you are giving 2 different actions:
1) Check the type of the variable and convert the variable to the correct data type
2) Perform a math calculation 

A function only needs to perform one action, so this problem can be solved with two functions.
The decorator takes the responsibility of converting the type of variables and passes the 'ready' data to the `add` function.